git-sync/main.go

```diff
@@ -134,40 +138,74 @@ func main() {
 	}
 }
 
-// syncRepo syncs the branch of a given repository to the destination at the given rev.
-func syncRepo(repo, dest, branch, rev string, depth int) error {
-	gitRepoPath := path.Join(dest, ".git")
-	_, err := os.Stat(gitRepoPath)
-	switch {
-	case os.IsNotExist(err):
-		// clone repo
-		args := []string{"clone", "--no-checkout", "-b", branch}
-		if depth != 0 {
-			args = append(args, "-depth")
-			args = append(args, string(depth))
+// updateSymlink atomically swaps the symlink to point at the specified directory and cleans up the previous worktree.
+func updateSymlink(link, newDir string) error {
+	// Get currently-linked repo directory (to be removed), unless it doesn't exist
+	currentDir, err := filepath.EvalSymlinks(path.Join(volMount, link))
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("error accessing symlink: %v", err)
+	}
+
+	// newDir is /git/rev-..., we need to change it to relative path.
+	// Volume in other container may not be mounted at /git, so the symlink can't point to /git.
+	newDirRelative, err := filepath.Rel(volMount, newDir)
+	if err != nil {
+		return fmt.Errorf("error converting to relative path: %v", err)
+	}
+
+	if _, err := runCommand("ln", volMount, []string{"-snf", newDirRelative, "tmp-link"}); err != nil {
```

Comment:
mv an ln can probably be done without exec a subprocess but we can leave for a followup